paho/pinger.go

```diff
@@ -81,106 +73,67 @@ func (p *DefaultPinger) Run(conn net.Conn, keepAlive uint16) error {
 		return fmt.Errorf("conn is nil")
 	}
 	p.mu.Lock()
-	if p.timer != nil {
+	if p.running {
 		p.mu.Unlock()
-		return fmt.Errorf("Run() already called")
+		return fmt.Errorf("Run() already in progress")
 	}
-	select {
-	case <-p.done:
+	p.running = true
+	p.mu.Unlock()
+	defer func() {
+		p.mu.Lock()
 		p.mu.Unlock()
```

Comment:
Guess defer is missing?

Reply:
Thanks, have moved the Unlock to the end.